### PR TITLE
Backtrace options through execinfo and gdb

### PR DESCRIPTION
**Pull request by wrrobin**
Related issue: #557 

Signed-off-by: Md <md.rahman@intel.com>

### CODE REVIEW

**Comment by wrrobin** [comment]
Sample configure output:
```
Compilers:
  C:              gcc
  C++:            g++ -std=gnu++11
  Fortran:        none

Transports:
  Portals 4:      no
  OFI:            yes

On Node Communication:
  XPMEM:          no
  CMA:            no
  memcpy (self):  no

Global Options:
  Remote VA:      no
  Thread support: yes
  Error checking: yes

OFI Transport:
  Memory reg.:    scalable
  Addr. vector:   table

Backtrace:        execinfo, gdb
```

**Comment by wrrobin** [comment]
Sample output from gdb
```
Using host libthread_db library "/usr/lib64/libthread_db.so.1".
0x00007f23b64092c9 in waitpid () from /usr/lib64/libc.so.6
0x00007f85a258c2c9 in waitpid () from /usr/lib64/libc.so.6
  Id   Target Id         Frame
  3    Thread 0x7f23b5400700 (LWP 153985) "hello" 0x00007f23b640966d in nanosleep () from /usr/lib64/libc.so.6
  2    Thread 0x7f23b383d700 (LWP 153990) "hello" 0x00007f23b6437dfd in poll () from /usr/lib64/libc.so.6
  Id   Target Id         Frame
  3    Thread 0x7f85a1583700 (LWP 153984) "hello" 0x00007f85a258c66d in nanosleep () from /usr/lib64/libc.so.6
* 1    Thread 0x7f23b6e4e740 (LWP 153983) "hello" 0x00007f23b64092c9 in waitpid () from /usr/lib64/libc.so.6
  2    Thread 0x7f859f9c0700 (LWP 153989) "hello" 0x00007f85a25badfd in poll () from /usr/lib64/libc.so.6
* 1    Thread 0x7f85a2fd1740 (LWP 153982) "hello" 0x00007f85a258c2c9 in waitpid () from /usr/lib64/libc.so.6

Thread 3 (Thread 0x7f23b5400700 (LWP 153985)):
#0  0x00007f23b640966d in nanosleep () from /usr/lib64/libc.so.6
#1  0x00007f23b60fedf4 in psmx2_progress_func () from /home/rahmanmd/sos-libfabric-psm2-check/libfabric-master/install/lib/libfabric.so.1

Thread 3 (Thread 0x7f85a1583700 (LWP 153984)):
#0  0x00007f85a258c66d in nanosleep () from /usr/lib64/libc.so.6
#1  0x00007f85a2281df4 in psmx2_progress_func () from /home/rahmanmd/sos-libfabric-psm2-check/libfabric-master/install/lib/libfabric.so.1
#2  0x00007f23b5a2fdc5 in start_thread () from /usr/lib64/libpthread.so.0
#3  0x00007f23b644273d in clone () from /usr/lib64/libc.so.6
#2  0x00007f85a1bb2dc5 in start_thread () from /usr/lib64/libpthread.so.0
#3  0x00007f85a25c573d in clone () from /usr/lib64/libc.so.6

Thread 2 (Thread 0x7f859f9c0700 (LWP 153989)):
#0  0x00007f85a25badfd in poll () from /usr/lib64/libc.so.6
#1  0x00007f85a1fef974 in ips_ptl_pollintr (rcvthreadc=0x19295d0) at /home/rahmanmd/intel-opa-psm2/opa-psm2/ptl_ips/ptl_rcvthread.c:379
#2  0x00007f85a1bb2dc5 in start_thread () from /usr/lib64/libpthread.so.0
#3  0x00007f85a25c573d in clone () from /usr/lib64/libc.so.6
```

**Comment by wrrobin** [comment]
@davidozog PR is ready for review.

**Comment by wrrobin** [comment]
Thanks @davidozog. Regarding artifact acknowledgment, I would leave that to @jdinan to decide.

File: src/runtime-pmix.c [hunk 1]
```diff
@@ -87,6 +87,10 @@ shmem_runtime_abort(int exit_code, const char msg[])
         __builtin_trap();
 #endif
 
+#ifdef USE_BT
+    collect_backtrace();
+#endif
+
```
[the remark this answers is not in the document]
Comment (wrrobin) [reply src/runtime-pmix.c]: Thanks for catching this. Not sure how and why I missed it.

File: src/backtrace.h [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+#else
+        fprintf(stderr, "Backtrace support through gdb is not available.\n");
+#endif
+    } else {
+        fprintf(stderr, "Ignoring bad user input for backtrace method %s. Choosing execinfo.\n",
+                       method);
+#if defined(USE_BT_EXECINFO)
+        backtrace_execinfo();
+#else
+        fprintf("Backtrace support through execinfo is not available.\n");
```
[the remark this answers is not in the document]
Comment (wrrobin) [reply src/backtrace.h]: I am not sure about this. If we add the other debugger options, then here we have to check all of them and use any one. I was thinking to use only `execinfo` by default, and if thats not available, then set the env parameter.

File: src/backtrace.h [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+static inline
+void collect_backtrace(void) {
+    char *method = shmem_internal_params.BACKTRACE_METHOD;
+
+    if (0 == strcmp(method, "execinfo")) {
+#if defined(USE_BT_EXECINFO)
+        backtrace_execinfo();
+#else
+        fprintf(stderr, "Backtrace support through execinfo is not available.\n");
```
[the remark this answers is not in the document]
Comment (wrrobin) [reply src/backtrace.h]: I just followed what `runtime-*.c` follows. I think the dependency on `shmem_internal.h` is also an issue here. I will check and see.

File: src/backtrace.h [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+#define MAX_BT_SIZE 1024
+
+static inline 
+void backtrace_execinfo(void) {
+    static void *btaddr[MAX_BT_SIZE];
+    size_t size, i;
+    char **fnnames = NULL;
+
+    size = backtrace(btaddr, MAX_BT_SIZE); 
+    fnnames = backtrace_symbols(btaddr, size);
```
[the remark this answers is not in the document]
Comment (wrrobin) [reply src/backtrace.h]: Good idea. Thanks.

File: configure.ac [hunk 5]
```diff
@@ -417,6 +417,41 @@ AM_CONDITIONAL([USE_PMI1], [test "$pmi_type" = "pmi1"])
...
+    use_bt_option="execinfo"
+    AC_DEFINE([USE_BT_EXECINFO], [1], [Execinfo library calls for backtrace])
+  elif test "$ac_cv_func_backtrace" != "yes" && test -n "$GDB_PATH"; then
+    use_bt_option="gdb"
+    AC_DEFINE([USE_BT_GDB], [1], [gdb for backtrace])
+  else
+    use_bt_option="execinfo, gdb"
+    AC_DEFINE([USE_BT_EXECINFO], [1], [Execinfo library calls for backtrace])
+    AC_DEFINE([USE_BT_GDB], [1], [gdb for backtrace])
+  fi
```
[the remark this answers is not in the document]
Comment (wrrobin) [reply configure.ac]: Yes, I have the same feeling. But, we could follow a similar approach to GASNet in such case, i.e. using config flag to enable each option. We will not need the env variable then.

File: configure.ac [hunk 6]
```diff
@@ -417,6 +417,41 @@ AM_CONDITIONAL([USE_PMI1], [test "$pmi_type" = "pmi1"])
...
+  CFLAGS="$CFLAGS -rdynamic"
```
[the remark this answers is not in the document]
Comment (wrrobin) [reply configure.ac]: Yes, you are right. I thought of checking this but eventually forgot. I will check whether we require this for all options.

File: src/backtrace.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (wrrobin) [reply src/backtrace.h]: @davidozog Having a second thought here. With the current code, `USE_BT_EXECINFO` is defined only when the backtrace symbols are present. So, we do not really have to check `HAVE_BACKTRACE_SYMBOLS` since the whole routine is already checked with `USE_BT_EXECINFO`. 
I have added the check for NULL for `fnnames` though. Let me know if you feel this is ok.

[the remark this answers is not in the document]
Comment (davidozog) [reply src/backtrace.h]: Ah, I see.  Maybe then this concern starts in the `configure.ac` [file here](https://github.com/Sandia-OpenSHMEM/SOS/pull/795/files#diff-67e997bcfdac55191033d57a16d1408aR438).

Should we be checking for `ac_cv_func_backtrace_symbols` in addition to `ac_cv_func_backtrace`?

[the remark this answers is not in the document]
Comment (wrrobin) [reply src/backtrace.h]: Thanks @davidozog. Yes, I added it now.

File: src/backtrace.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidozog) [reply src/backtrace.h]: Yeah, I think you're probably right.  I'm a little surprised the `runtime-*.c` files aren't using the `RAISE_ERROR_*` routines... but there's possibly a good reason for that.  Perhaps SOS may not yet be initialized in the runtime layer, so shouldn't call `shmem_runtime_abort` via the `RAISE/RETURN` macros.  If that's true, it seems like a good enough reason for the backtrace to print to `stderr` as well).

File: src/backtrace.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidozog) [reply src/backtrace.h]: Yeah, sounds good.

**Comment by davidozog** [comment]
Might we want to cover the new code in the Travis tests?

**Comment by davidozog** [comment]
Do we need the license in `backtrace.h` as well?

**Comment by wrrobin** [comment]
Thanks @davidozog. Just added. Please let me know if this looks ok.

**Comment by davidozog** [comment]
Thanks @wrrobin, looks good.

That's it for my review.  Does @jdinan want to have a look before merging?

**Comment by jdinan** [comment]
Is a configure flag the right way to go?  Can we detect support at configure time and allow the user to enable this at runtime?  That is, default for `SHMEM_BACKTRACE` is the empty string (none) and user can set to `yes` (automatically pick which one to use), `gdb`, or `execinfo`?  Seems like we should also have something `SHMEM_BACKTRACE_PATH` to allow users to specify the path to gcc?

**Comment by wrrobin** [comment]
@jdinan Not sure if I am understanding the question. The current approach is to detect both the possible options (execinfo, gdb) at configure time and user can enable any one through `SHMEM_BACKTRACE_METHOD`. By default, if `SHMEM_BACKTRACE_METHOD` would have execinfo if that is available. Are you asking if the user has a custom location for execinfo, then how he can enable that? In such case, yes, we might have additional parameters for users to enable.

**Comment by davidozog** [comment]
I think @jdinan is saying that we could do the check for gdb/execinfo by default (i.e. without requiring `--enable-backtrace`), then turn it on with an environment variable (and if no backtracing tools are available, then perhaps print something like "Sorry, backtrace is not available"?).  This is sorta how `SHMEM_DEBUG` works.  I do think that would be more user-friendly.  For example, I'd personally rather not have to add this configure flag to my already too long list of configure flags... And from now on, I think most developers would always enable it since it's not on the critical path (right?) and it's so incredibly useful.

**Comment by wrrobin** [comment]
Ahh. yes.. that definitely sounds more user-friendly. Sorry that I did not get it first time. I will then make this change.

File: src/backtrace.h [hunk 7]
```diff
@@ -191,6 +191,14 @@ void collect_backtrace(void) {
         backtrace_gdb();
 #else
         fprintf(stderr, "Backtrace support through gdb is not available.\n");
+#endif
+    } else if (0 == strcmp(method, "auto")) {
```
Comment (davidozog) [inline src/backtrace.h]: I wonder here if we should support "yes" or any other string(s) here.  If a user sets BACKTRACE_METHOD=1, or anything that's not "execinfo" or "gdb", should we just fall into the "auto" case?  This is just a thought, it might be undesirable/overkill.

Comment (wrrobin) [reply src/backtrace.h]: @davidozog If the user sets anything other than "execinfo"/"gdb"/"auto", the "else" block will be executed and it will check for execinfo. If execinfo is not available, then it is exiting with a message. This behavior is slightly different than falling into "auto" (gdb will not be checked). Let me know if you guys think "auto" should be better option for invalid user input.

Comment (davidozog) [reply src/backtrace.h]: Sounds good, I'm ok with either way too.

It would be good to clarify in the README that "auto" is the valid input for (i.e. "Options are : auto, execinfo, gdb").  I think some vars use the "(default: auto)" description where "auto" is not actually the input string (like `SHMEM_OFI_PROVIDER`).

Comment (wrrobin) [reply src/backtrace.h]: Right. Thanks for noting that. I was also thinking whether we should have something on backtracing in the Wiki. As we add more methods in the future, I think it will be better if we have a small description on the Wiki page.

File: src/shmem_env_defs.h [hunk 8]
```diff
@@ -105,7 +105,5 @@ SHMEM_INTERNAL_ENV_DEF(MPI_THREAD_LEVEL, string, "MPI_THREAD_SINGLE", SHMEM_INTE
                        "Specify the MPI threading level when MPI is used as the process manager")
 #endif
 
-#ifdef USE_BT
-SHMEM_INTERNAL_ENV_DEF(BACKTRACE_METHOD, string, "execinfo", SHMEM_INTERNAL_ENV_CAT_OTHER, 
+SHMEM_INTERNAL_ENV_DEF(BACKTRACE_METHOD, string, "auto", SHMEM_INTERNAL_ENV_CAT_OTHER,
```
[the remark this answers is not in the document]
Comment (wrrobin) [reply src/shmem_env_defs.h]: I think shortening this to `SHMEM_BACKTRACE` sounds like an env variable used to enable/disable backtrace. However, we are trying to use this variable for choosing the method user wants, right? 

If we choose "disabled" the default, then user will see no backtraces unless they set this parameter. I can see then why you would like `SHMEM_BACKTRACE` as opposed to `SHMEM_BACKTRACE_METHOD`. We can go with this then if you guys think this would be ideal.

File: configure.ac [hunk 9]
```diff
@@ -418,44 +418,41 @@ AM_CONDITIONAL([USE_PMI2], [test "$pmi_type" = "pmi2"])
...
+AC_PATH_PROGS(GDB_PATH, $GDB gdb, , /usr/bin:${PATH})
+if test -n "$GDB_PATH"; then
+  AC_DEFINE_UNQUOTED(GDB_PATH,"$GDB_PATH")
+else
+  AC_DEFINE([GDB_PATH], [""], [Path to gdb location])
+fi
+
+if test "$bt_execinfo" != "yes" && test -z "$GDB_PATH"; then
+  use_bt_option="N/A"
+  AC_MSG_ERROR(Backtrace cannot be provided as support for execinfo and gdb is not found in the system)
```
[the remark this answers is not in the document]
Comment (wrrobin) [reply configure.ac]: I was thinking to keep it like this for this PR. As soon as we bring lldb, we are changing these checks anyway. Omitting the error is also fine, I guess. We can just leave a comment for future ref.

File: src/backtrace.h [hunk 10]
```diff
@@ -191,6 +191,14 @@ void collect_backtrace(void) {
...
+#if defined(USE_BT_EXECINFO)
+        backtrace_execinfo();
+#elif defined(USE_BT_GDB)
+        backtrace_gdb();
+#else
+        fprintf(stderr, "Backtrace support is not available.\n");
```
[the remark this answers is not in the document]
Comment (wrrobin) [reply src/backtrace.h]: @jdinan In this condition, user requested "auto", but none of the two options are available. So, we should print this message, right?

[the remark this answers is not in the document]
Comment (wrrobin) [reply src/backtrace.h]: @jdinan There is a cyclic dependency here if `RAISE_WARN_STR` macros are used. `shmem_internal.h` includes `runtime.h` which in turn includes `backtrace.h`. Now, to use `RAISE_WARN_STR` in `backtrace.h`, we will have to include `shmem_internal.h` in `backtrace.h`. Let me know if you see a work-around for this.

File: configure.ac [hunk 11]
```diff
@@ -418,44 +418,41 @@ AM_CONDITIONAL([USE_PMI2], [test "$pmi_type" = "pmi2"])
...
-  if test -n "$GDB_PATH"; then
-    AC_DEFINE_UNQUOTED(GDB_PATH,"$GDB_PATH")
-  else
-    AC_DEFINE([GDB_PATH], [""], [Path to gdb location])
+## execinfo
+AC_CHECK_HEADERS(execinfo.h)
+bt_execinfo="no"
+if test "$ac_cv_header_execinfo_h" = "yes" ; then
+  AC_SEARCH_LIBS(backtrace, execinfo) 
+  AC_CHECK_FUNCS(backtrace backtrace_symbols)
```
[the remark this answers is not in the document]
Comment (wrrobin) [reply configure.ac]: Oh, ok. I will add it then.

File: src/backtrace.h [hunk 12]
```diff
@@ -200,6 +206,8 @@ void collect_backtrace(void) {
 #else
         fprintf(stderr, "Backtrace support is not available.\n");        
 #endif
+    } else if (0 == strcmp(method, "")) {
+        fprintf(stderr, "Backtrace support is not enabled. Set SHMEM_BACKTRACE to auto, execinfo, or gdb to enable.\n");
```
Comment (jdinan) [inline src/backtrace.h]: Better not to list acceptable values here, since it adds another place where we need to maintain this list.  I would shorten this to: `Set SHMEM_BACKTRACE for additional details`.  But, also, is it necessary to advertise the feature in this way?

File: README [hunk 13]
```diff
@@ -249,7 +249,11 @@ options.
...
         can be used to interface with a debugger or generate core files.
 
-    SHMEM_BACKTRACE_METHOD (default: auto)
-        Can be used to choose the backtracing mechanism. By default, it chooses  
-        any one of the available options between Linux library, execinfo and 
-        external debugging tool, gdb.  
+    SHMEM_BACKTRACE (default: <empty>)
+        Can be used to choose the backtracing mechanism. User can set this with 
+        any one of the available options among Linux library, execinfo and 
+        external debugging tool, gdb, or auto.
```
Comment (jdinan) [inline README]: Please give a list of the exact values that are accepted by this env var.

File: src/backtrace.h [hunk 14]
```diff
@@ -150,7 +150,13 @@ void backtrace_gdb(void) {
...
     static char cmd[sizeof(fmt) + 3 * MAX_BT_PATHSIZE];
     char filename[MAX_BT_PATHSIZE];
-    const char *gdb = (access(GDB_PATH, X_OK) ? "gdb" : GDB_PATH);
+    char *gdb = NULL;
+    char *backtrace_path = shmem_internal_params.BACKTRACE_PATH;
+    if (0 == strcmp(backtrace_path, "")) {
+        gdb = (access(GDB_PATH, X_OK) ? "gdb" : GDB_PATH);
+    } else {
+        gdb = (access(backtrace_path, X_OK) ? "gdb" : backtrace_path);
+    }
```
[the remark this answers is not in the document]
Comment (wrrobin) [reply src/backtrace.h]: No, these just get the URL based on the paths provided. If the permission does not allow from `access` then just `gdb` is added as the URL since we already defined `USE_BT_GDB`.

File: src/backtrace.h [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+static inline 
+void backtrace_gdb(void) {
+    const char fmt[] = "%s -nx -batch -x %s '%s' %d";
+    static char cmd[sizeof(fmt) + 3 * MAX_BT_PATHSIZE];
+    char filename[MAX_BT_PATHSIZE];
+    char *gdb = NULL;
+    char *backtrace_path = shmem_internal_params.BACKTRACE_PATH;
+    if (0 == strcmp(backtrace_path, "")) {
```
Comment (jdinan) [inline src/backtrace.h]: Use `shmem_internal_params.BACKTRACE_PATH_provided` for this check

**Comment by jdinan** [comment]
With regard to advertising the backtrace feature, I'm not sure we always want this.  For example, some programs call `shmem_global_exit` with exit status `0` to exit successfully.
```
$ mpiexec -np 2 test/unit/global_exit
[0000] WARN:  ../../src/backtrace.c:215: shmem_util_backtrace
[0000]        Backtrace support is not enabled. See SHMEM_BACKTRACE for additional details.
```

**Comment by jdinan** [comment]
I think I prefer to always build GDB support and do a runtime check for the debugger.  Will look into this...

**Comment by wrrobin** [comment]
@jdinan, do you mean using GDB as the first option for "auto" choice?

File: src/backtrace.c [hunk 16]
```diff
@@ -150,8 +148,12 @@ void backtrace_gdb(void) {
     const char fmt[] = "%s -nx -batch -x %s -p %d";
     static char cmd[sizeof(fmt) + 3 * MAX_BT_PATHSIZE];
     char filename[MAX_BT_PATHSIZE];
-    const char *gdb = (access(GDB_PATH, X_OK) ? "gdb" : GDB_PATH);
     int rc;
+#ifdef GDB_PATH
+    const char *gdb = (access(GDB_PATH, X_OK) ? "gdb" : GDB_PATH);
+#else
+    const char *gdb = "gdb";
```
Comment (wrrobin) [inline src/backtrace.c]: @jdinan In this case, the system path might not have the "gdb" in it. Still executing with "gdb" will result in an error. Are we still going to proceed with this? Maybe I am not getting something here?

File: src/backtrace.c [hunk 17]
```diff
@@ -209,10 +209,10 @@ void shmem_util_backtrace(void) {
 #elif defined(USE_BT_GDB)
         backtrace_gdb();
 #else
-        RAISE_WARN_STR("Backtrace support is not available.");        
+        RAISE_WARN_STR("Backtrace support is not available.");
 #endif
     } else if (0 == strcmp(method, "")) {
-        RAISE_WARN_STR("Backtrace support is not enabled. See SHMEM_BACKTRACE for additional details.");        
+        RAISE_WARN_STR("Backtrace support is not enabled. See SHMEM_BACKTRACE for additional details.");
     } else {
```
Comment (wrrobin) [inline src/backtrace.c]: Sorry about the extra whitespaces. Thanks for the fixes.

File: src/backtrace.c (hunk 16 above)
Comment (jdinan) [reply src/backtrace.c]: That's what I was thinking.  We could write some code to check if GDB is in the path.  Otherwise, you get something like this:
```
$ SHMEM_BACKTRACE=gdb test/unit/global_exit
sh: gdb: command not found
[0000] ERROR: ../../src/backtrace.c:175: backtrace_gdb
[0000]        Error generating backtrace using gdb (32512)
PE 0 called shmem_global_exit with status 0
```

Comment (wrrobin) [reply src/backtrace.c]: Ok. But, we are already checking that in the configure, right? If it is not in the path, we were not defining "GDB_PATH"

**Comment by wrrobin** [comment]
@jdinan Let me know if you are ok with merging this.

File: src/backtrace.c (hunk 16 above)
Comment (jdinan) [reply src/backtrace.c]: Yeah, there is definitely overlap.  If I had to pick one of these, I think I would choose to use the runtime path to find gdb rather than what was found by configure

**Comment by jdinan** [comment]
@wrrobin Yes, I think so, as long as you're ok with the path discussion.

File: src/backtrace.c (hunk 16 above)
Comment (wrrobin) [reply src/backtrace.c]: Ok. Will take a look at this as a separate issue.
